BaseClasses.py

```diff
@@ -754,20 +737,116 @@ def can_reach_entrance(self, spot: str, player: int) -> bool:
     def can_reach_region(self, spot: str, player: int) -> bool:
         return self.multiworld.get_region(spot, player).can_reach(self)
 
-    def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> None:
+    def _sweep_for_events_impl(self, events_per_player: List[Tuple[int, List[Location]]], yield_each_sweep: bool,
+                               ) -> Iterable[None]:
+        """
+        The implementation for sweep_for_events is separated here because it returns a generator due to the use of a
+        yield statement.
+        """
+        all_players = {player for player, _ in events_per_player}
+        players_to_check = all_players
+        # As an optimization, it is assumed that each player's world only logically depends on itself. However, worlds
+        # are allowed to logically depend on other worlds, so once there are no more players that should be checked
+        # under this assumption, an extra sweep iteration is performed that checks every player, to confirm that the
+        # sweep is finished.
+        checking_if_finished = False
+        while players_to_check:
+            next_events_per_player: List[Tuple[int, List[Location]]] = []
+            next_players_to_check = set()
+
+            for player, locations in events_per_player:
+                if player not in players_to_check:
+                    next_events_per_player.append((player, locations))
+                    continue
+
+                # Accessibility of each location is checked first because a player's region accessibility cache becomes
+                # stale whenever one of their own items is collected into the state.
+                reachable_locations: List[Location] = []
+                unreachable_locations: List[Location] = []
+                for location in locations:
+                    if location.can_reach(self):
+                        # Locations containing items that do not belong to `player` could be collected immediately
+                        # because they won't stale `player`'s region accessibility cache, but, for simplicity, all the
+                        # items at reachable locations are collected in a single loop.
+                        reachable_locations.append(location)
+                    else:
+                        unreachable_locations.append(location)
+                if unreachable_locations:
+                    next_events_per_player.append((player, unreachable_locations))
+
+                # If the player does not receive any items for the rest of the sweep iteration, they will likely not
+                # have any reachable locations at the start of the next sweep iteration, so their locations can be
+                # skipped.
+                next_players_to_check.discard(player)
```

Comment:
I did some testing here on a hunch that `.discard` and `.add` on a set are slower than `dict[player] = True` and `dict[player] = False`.

This turned out to be true, but the useful thing you gain by using a set that you *remove* from is that you can **iterate it directly** and also check for its length to know when you're done.

The added overhead of checking the True or False from the dict for each player makes it slower.

So what you did is probably best